### PR TITLE
Add Bitcoin Core RPC cookie authentication option

### LINKED ISSUE #2806: Add support for rpccookiefile in bitcoind.conf
### Description

In order to avoid username/passwords which are not the recommended config for bitcoind it would be great if i could import the rpccookiefile for bitcoind instead of using user/pass

### Problem to be solved

lnd, electrs, and most other tools i use utilize the cookie file to avoid using a static user:pass, but I'm unable to use mempool with this configuration because it appears as though there is no support for the cookie file

### Proposed solution

allow mempool to use the cookie file for bitcoind rpc authentication

### PR DESCRIPTION
**Pull request by antonilol**
still a draft, i know from #1180 that config changes also need changes in docker related files

now setting `COOKIE` to true should be enough on most systems, where bitcoind places the cookie in the expected (network specific) dir (`~/.bitcoin/<network>/.cookie`), otherwise `COOKIE_PATH` will point mempool in the right direction

user/pass auth is still the default though

the default path of the cookie is not set using the normal way (like every other config entry) because it is network specific, it looks like a dirty workaround.

i also refractored the `rpc-api` a bit, this helped me and i think this will also help other working on this project, typescript can now tell you if the rpc name is correct (the diff of jsonrpc.ts is unreadable because it ended up one level of indentation deeper)

@ShadwDrgn @federicociro @BeardMatt (from the linked issue) can you review my choices/code or test it and tell me if this suits your needs?

feedback/tips on how to improve this are all appreciated!

fixes #2806

### CODE REVIEW

**Comment by ShadwDrgn** [comment]
haven't tested but based on what i'm  r eading it looks like we just add COOKIE and COOKIE_PATH variables to our compose environment section as needed and ensure the cookie file is mounted. should work for my purposes at least

**Comment by federicociro** [comment]
Will do a review of the change after my holidays

**Review by federicociro — COMMENTED** [review]
Change approved and performed as expected. Mempool backend connects to bitcoind correctly through .cookie authentication.

**Comment by antonilol** [comment]
just squashed and rebased this, happy new year everyone! do not forget to update your open source licenses' dates :)

**Review by federicociro — COMMENTED** [review]
How can we push this? Everything is ok.

**Comment by antonilol** [comment]
this needs review from a maintainer
there is a branch conflict, i will rebase in a second

**Review by wiz — CHANGES_REQUESTED** [review]
Concept ACK for adding cookie support, but Concept NACK for rewriting the entire RPC code... most importantly, that is an unrelated change that should not be part of this PR

**Comment by antonilol** [comment]
i separated the two changes in two separate prs, the second pr contains the start of #2840 as this is not necessary for cookie auth

File: backend/mempool-config.sample.json
Comment (antonilol) [inline backend/mempool-config.sample.json:38]: it would be great to have cookie work by default, but setting cookie to true here will turn it on for everybody that did not set in in his config, potentially breaking mempool if the cookie is not in the path set here, note that this is also different for different networks (mainnet cookie is in .bitcoin/.cookie, but testnet in .bitcoin/testnet3/.cookie, and not everyone has his .bitcoin in the home directory)

i think i can change the config and sample config merging to allow for some special logic to use username and password if that is configured, cookie if that is configured and the default cookie path for that network if no auth method is set in the config, or maybe even try some known .bitcoin locations to try to find it (some of these full node distros have it on /mnt for example)

File: backend/src/api/bitcoin/bitcoin-client.ts
Comment (wiz) [inline backend/src/api/bitcoin/bitcoin-client.ts:6]: I don't think we want to do this, you can configure your cookie path in the configuration file and the default should be `/bitcoin/.cookie`

Comment (antonilol) [reply backend/src/api/bitcoin/bitcoin-client.ts:6]: this is the cookie location on my device, i run bitcoind and mempool from the same unix account without docker

on which systems is `/bitcoin` the bitcoind data dir?

**Comment by ShadwDrgn** [comment]
I'm curious if this is just being held up about confusion as to where the bitcoin configs are/should be? This merge seems like it's been ready for some time now. Anything I or anyone else can do to help?

File: backend/src/api/bitcoin/bitcoin-client.ts
Comment (wiz) [reply backend/src/api/bitcoin/bitcoin-client.ts:6]: the official mempool.space servers

Comment (antonilol) [reply backend/src/api/bitcoin/bitcoin-client.ts:6]: ok, changing this

**Comment by antonilol** [comment]
can this be reviewed/merged?

**Comment by knorrium** [comment]
The tests are failing due to missing docker config keys and script update

```
Mempool Backend Config › should ensure that the mempool-config.json Docker template has all the keys

    expect(received).toContain(expected) // indexOf

    Expected substring: "\"COOKIE\": \"__CORE_RPC_COOKIE__\""
```

```
 ● Mempool Backend Config › should ensure the docker start.sh script has default values

    expect(received).toMatch(expected)

    Expected pattern: /__CORE_RPC_COOKIE_PATH__=\${CORE_RPC_COOKIE_PATH:=(.*)}/
```

**Comment by antonilol** [comment]
fixed and tests pass now

**Review by softsimon — APPROVED** [review]
tested ACK @ [06d1bbdf0]

Tested using with and no cookie and with and without on the second core node.

**Pull request merged**
